### ISSUE TITLE
On-the-fly send to Kindle failing

### ISSUE THREAD

**Issue by malletfils** [janeczku/calibre-web#1794]
Hello,

The "Send to Kindle" function fails for Epub-only ebooks when there is a single quote in their title.

The manual conversion from Calibre-Web from epub to mobi still works though, so it is possible to convert the book and then use "Send to Kindle" function. But if you use "Send to Kindle" on non-previously converted books (and only when they have a single quote in their titles), it fails, and then you have to restart Calibre-web (if you launch other tasks after the failure, they keep waiting in line).

It shows that on log :

[2021-01-02 18:56:55,881]  INFO {cps:100} Starting Calibre Web...
[2021-01-02 18:56:55,983]  INFO {cps.server:168} Starting Tornado server on :8083
[2021-01-02 18:58:27,620] ERROR {cps:1892} Exception on /admin/logdownload/0 [GET]
Traceback (most recent call last):
  File "/home/zejedi/calibre-web/vendor/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/zejedi/calibre-web/vendor/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/zejedi/calibre-web/vendor/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/zejedi/calibre-web/vendor/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/zejedi/calibre-web/vendor/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/zejedi/calibre-web/vendor/flask_login/utils.py", line 272, in decorated_view
    return func(*args, **kwargs)
  File "/home/zejedi/calibre-web/cps/admin.py", line 90, in inner
    return f(*args, **kwargs)
  File "/home/zejedi/calibre-web/cps/admin.py", line 1287, in download_log
    return debug_info.assemble_logfiles(file_name)
  File "/home/zejedi/calibre-web/cps/debug_info.py", line 39, in assemble_logfiles
    shutil.copyfileobj(fd, wfd)
  File "/usr/lib/python2.7/shutil.py", line 66, in copyfileobj
    fdst.write(buf)
TypeError: unicode argument expected, got 'str'

**Comment by OzzieIsaacs** [comment]
I fixed the export of the logfile with python2. I tried to do on the fly conversion for a epub book with ' in the title -> `GRIP'o`, but everything was perfect.
What exactly fails, the conversion, or the sending of the book?
Could you please repeat your errorcase with loglevel set to "debug" and send me the logfile. Preferably the file you get by pushing the `Download Debug Package` button

**Comment by malletfils** [comment]
Hello,

I've just updated Calibre-Web to fix the log problem, set my log level to debug, and made multiple tries with books. You are right, this is not just the presence of the single quote that causes the crash, because for other titles, it works. See below :

zejedi	Convert: epub -> mobi: L'Illusion	Finished	100 %	15s	1/3/21, 4:54 PM
zejedi	Email: Send to Kindle	Finished	100 %	0s	1/3/21, 4:54 PM
zejedi	Convert: epub -> mobi: Le Crépuscule et l'Aube	Failed	100 %	2:55s	1/3/21, 4:54 PM

But I still have nothing in the logs (they end with the previous conversion). I can send them to you (but how ? I don't see any option to send a file).

It fails at the conversion step (I see no mobi file created on the server). It could be for this specific book, but it works fine when I convert it manually : 

zejedi | Convert: EPUB -> MOBI: Le Crépuscule et l'Aube | Finished | 100 % | 37s | 1/3/21, 5:12 PM
zejedi | Convert: epub -> mobi: Les Étoiles s’éteignent à l’aube | Finished | 100 % | 6s | 1/3/21, 5:14 PM
zejedi | Email: Send to Kindle | Finished | 100 % | 0s | 1/3/21, 5:14 PM
zejedi | Convert: epub -> mobi: Tenir jusqu'à l'Aube | Failed | 100 % | 25s | 1/3/21, 5:15 PM

Then I just restart the server, and retry "Send to Kindle", and it works (just restarting was not enough though for my 1st example)

zejedi | Convert: epub -> mobi: Tenir jusqu'à l'Aube | Finished | 100 % | 6s | 1/3/21, 5:16 PM
zejedi | Email: Send to Kindle | Finished | 100 % | 0s | 1/3/21, 5:16 PM

**Comment by OzzieIsaacs** [comment]
Please send it to me via private email (on the GitHub profile page).
Please send me the debug package file

**Comment by OzzieIsaacs** [comment]
Some bad news: I made a mistake, so your email password is in the logfile, please change it.
What I can read from the information you provided me:
I saw you are using Ubuntu (I guess 20.04), if so, please consider updating to python3.x as the python2 version is out of support for more than one year and I'm only testing calibre-web with python3, and it looks like logbook output from the backgrund conversion is not very reliable. 
If my guess it true and you are using ubuntu 20.04 please also consider updating `calibre` to the newest version 5.8.2 instead of 4.22, as it looks the problem is somehow related to the calibre-converter itself. (This line in the logbook is a calibre ebook convert error-message`[2021-01-03 16:54:16,063] DEBUG {cps.tasks.convert:216} QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-zejedi'`). 
Why this is only affecting a combined convert and send task I can't explain now.
My guess it's the book you are trying to convert has a problem, you could try to use Sigil (https://github.com/Sigil-Ebook/Sigil/releases) to check the epub file. The last time I used it there were some buttons to detect and fix epub problems.

The failed conversion are indeed nowhere to see and you also don't see the last restart. Strange.

I you could send me a problematic book (via email) I could try to reproduce the problem on my computer.

**Comment by malletfils** [comment]
Hello,

Thanks for your support. OK for the password, I will change it. Not a big trouble, because this is an email adress only dedicated to this function. I am on Ubuntu 20.04, with Python3 installed. I just saw that the 2.7 version was used instead of the 3.8.5. Changed that. And updated Calibre too (as this version is only used by Calibre-Web, I don't think about updating it often. That would be an idea to put some warning somewhere in Calibre-Web when the calibre version is not up to date).

I can't seem to be able to reproduce the problem anymore.  So solved for me.

All the best for the new year, and thanks again for your wonderful work (and lightning fast support :) ) !
